### ISSUE TITLE
Incorrect assign to integer64 type

### ISSUE THREAD

**Issue by artemklevtsov** [Rdatatable/data.table#2530]
Code to reproduce.

``` r
suppressMessages(library(data.table))
suppressMessages(library(bit64))
set.seed(0)
DT <- data.table(
    a = as.integer64(1:10)
)
DT[a == 1, a := 12]
DT[a == 2, a := as.integer64(12)]
print(DT)
#>                       a
#>  1: 4622945017495814144
#>  2:                  12
#>  3:                   3
#>  4:                   4
#>  5:                   5
#>  6:                   6
#>  7:                   7
#>  8:                   8
#>  9:                   9
#> 10:                  10
```

<details>
<summary>Session info</summary>

``` r
devtools::session_info()
#> Session info -------------------------------------------------------------
#>  setting  value                       
#>  version  R version 3.4.3 (2017-11-30)
#>  system   x86_64, linux-gnu           
#>  ui       X11                         
#>  language ru                          
#>  collate  ru_RU.UTF-8                 
#>  tz       Asia/Novosibirsk            
#>  date     2017-12-17
#> Packages -----------------------------------------------------------------
#>  package    * version  date       source        
#>  backports    1.1.1    2017-09-25 CRAN (R 3.4.2)
#>  base       * 3.4.3    2017-11-30 local         
#>  bit        * 1.1-12   2014-04-09 CRAN (R 3.4.0)
#>  bit64      * 0.9-7    2017-05-08 CRAN (R 3.4.0)
#>  compiler     3.4.3    2017-11-30 local         
#>  data.table * 1.10.4-3 2017-10-27 CRAN (R 3.4.3)
#>  datasets   * 3.4.3    2017-11-30 local         
#>  devtools     1.13.4   2017-11-09 CRAN (R 3.4.2)
#>  digest       0.6.12   2017-01-27 CRAN (R 3.4.0)
#>  evaluate     0.10.1   2017-06-24 CRAN (R 3.4.0)
#>  graphics   * 3.4.3    2017-11-30 local         
#>  grDevices  * 3.4.3    2017-11-30 local         
#>  htmltools    0.3.6    2017-04-28 CRAN (R 3.4.0)
#>  knitr        1.17     2017-08-10 CRAN (R 3.4.1)
#>  magrittr     1.5      2014-11-22 CRAN (R 3.4.0)
#>  memoise      1.1.0    2017-04-21 CRAN (R 3.4.0)
#>  methods    * 3.4.3    2017-11-30 local         
#>  Rcpp         0.12.14  2017-11-23 CRAN (R 3.4.2)
#>  rmarkdown    1.8      2017-11-17 CRAN (R 3.4.2)
#>  rprojroot    1.2      2017-01-16 CRAN (R 3.4.0)
#>  stats      * 3.4.3    2017-11-30 local         
#>  stringi      1.1.6    2017-11-17 CRAN (R 3.4.2)
#>  stringr      1.2.0    2017-02-18 CRAN (R 3.4.0)
#>  tools        3.4.3    2017-11-30 local         
#>  utils      * 3.4.3    2017-11-30 local         
#>  withr        2.1.0    2017-11-01 CRAN (R 3.4.2)
#>  yaml         2.1.16   2017-12-12 CRAN (R 3.4.3)
```

</details>

**Comment by franknarf1** [comment]
Looks like correct behavior..?

```
library(bit64)
structure(12, class="integer64")
# integer64
 # [1] 4622945017495814144
```

I think as long as the storage mode between lhs and rhs matches, it will assign without coercion. 

Seems arguably analogous to ...

```
DT = data.table(id = 1:2, d = Sys.Date())
DT[id == 1, d := 2011-01-01]
DT[id == 2, d := as.Date("2011-01-01") ][]
#    id          d
# 1:  1 1975-07-03
# 2:  2 2011-01-01
```

Maybe it would be nice to have an option to strictly enforce class and attribute matching when subassigning with `:=`, though.

**Comment by mattdowle** [comment]
Thanks, @artemklevtsov. It seems this was fixed in v1.12.4 (Oct 2019) very likely by new feature 29 introduced then.

![image](https://user-images.githubusercontent.com/2779998/121609671-98b22a80-ca11-11eb-8927-1680c4d69c34.png)

Thanks @jangorecki for adding your test in PR #4578 to make sure it doesn't reoccur.
